### ISSUE TITLE
Support installation in read-only locations

### ISSUE THREAD

**Issue by fedeinthemix** [non-Jedi/eglot-jl#53]
At the moment `eglot-jl` can't be installed as a system package as it needs to write to the package directory. It appears that the new Julia 1.11 feature [Precompile file relocatability](https://julialang.org/blog/2024/10/julia-1.11-highlights/#precompile_file_relocatability) may help.

**Comment by non-Jedi** [comment]
I'm a bit confused here. Are you talking about writing to the emacs package directory or to the Julia depot? As far as I know, eglot-jl shouldn't need to write to the directory in which its installed.

Not writing to the Julia depot is out of scope since installing the language server automatically is one of the selling-points of this  package. If you want to manage the language server installation yourself, you're better off doing so without this package.

**Comment by non-Jedi** [comment]
Does #54 solve the problem you're having?

**Comment by fedeinthemix** [comment]
> I'm a bit confused here. Are you talking about writing to the emacs package directory or to the Julia depot? As far as I know, eglot-jl shouldn't need to write to the directory in which its installed.
> 
> Not writing to the Julia depot is out of scope since installing the language server automatically is one of the selling-points of this package. If you want to manage the language server installation yourself, you're better off doing so without this package.

I'm using NixOS which builds all packages in isolated sandboxes and puts the resulting artifacts in read-only locations (for security and reproducibility reasons). Trying to install `eglot-jl` fails, while if I install it as a user it works.

If I look into the package installed in `~/.emacs.d/elpa/eglot-jl-20240911.1352`, I see that `Manifest.toml` has a date later than the installation date. I guess that's the (first?) problem.

The sandbox doesn't have access to the user depot, but I can designate an immutable directory where to put the files it needs. Is this feasible somehow?

**Comment by non-Jedi** [comment]
> If I look into the package installed in `~/.emacs.d/elpa/eglot-jl-20240911.1352`, I see that `Manifest.toml` has a date later than the installation date. I guess that's the (first?) problem.

If you're using Julia 1.11, could you please try out #54? I believe it solves this problem although there are some questions to be answered about how to handle patch releases.

If this doesn't work, in general, what you may want to do is include a build step in your nix expression(?) that simply runs `julia --project=/path/to/eglot-jl/build/dir /path/to/eglot-jl/build/dir/eglot-jl.jl /path/to/eglot-jl/build/dir/` and kills the process after about two minutes.

Any update to the Manifest should only happen once when a new version of Julia is used. This approach does have the disadvantage that it will require a revbump to eglot-jl on each update to Julia, but I'm not seeing a way around that right now with Julia's Manifest design.

**Comment by fedeinthemix** [comment]
> If you're using Julia 1.11, could you please try out #54? I believe it solves this problem although there are some questions to be answered about how to handle patch releases.

I'm following the stable channel which provides Julia 1.10.3. I'll try when Julia 1.11 will come available.

> 
> If this doesn't work, in general, what you may want to do is include a build step in your nix expression(?) that simply runs `julia --project=/path/to/eglot-jl/build/dir /path/to/eglot-jl/build/dir/eglot-jl.jl /path/to/eglot-jl/build/dir/` and kills the process after about two minutes.

I'll try something along these lines, but killing the process after some time is an ugly hack.

> Any update to the Manifest should only happen once when a new version of Julia is used. This approach does have the disadvantage that it will require a revbump to eglot-jl on each update to Julia, but I'm not seeing a way around that right now with Julia's Manifest design.

In NixOS everything is updated together which means that when Julia is updated, eglot-jl will be rebuilt. So, this should not be a problem.

**Comment by non-Jedi** [comment]
> I'll try something along these lines, but killing the process after some time is an ugly hack.

You can try just running `Pkg.resolve()` in the eglot-jl environment. That should be sufficient for the Manifest problem at least, but I'm still not sure how the Julia depot will work in your setup.

**Comment by tpapp** [comment]
Sorry for the stupid question, but if the `Manifest.toml` is regenerated anyway then does it need to be part of the repository?

I am facing the same problem to a certain extent (straight.el does not like when the repo has uncommitted changes and won't update after that).

**Comment by non-Jedi** [comment]
This whole philosophical thing about whether Manifests should be committed or not is probably worth another forum post to hash things out. The position that I remember everyone settling on when Pkg3 was first coming into use was that Manifests should be committed for "applications" to ensure the app would have deterministic dependencies but that for libraries it shouldn't be. `eglot-jl` is definitely an application.

The problem is that the Manifest encodes the specific julia version used and can even change in breaking ways between Julia versions. So there's currently no way to package something with pinned dependencies other than Julia itself other than just manually pinning everything in Project.toml.

So all my options kind of stink for this one:

1. Current approach: include a Manifest, but try to update it if something is broken.
2. Include a Manifest, but if something is broken (e.g. Julia is a different version than I last updated the package with), let the user deal with the error himself.
3. Don't include a Manifest, and let Pkg grab whatever dependency versions it will when the environment is instantiated.
4. Don't include a Manifest but manually pin all dependency versions myself in Project.toml (obnoxious for maintainer)
5. Something else I haven't thought of?

**Comment by non-Jedi** [comment]
> Sorry for the stupid question, but if the `Manifest.toml` is regenerated anyway then does it need to be part of the repository?

I believe that the Manifest is only regenerated if it's out of sync (usually because the Manifest was generated with a different version of julia), but the docstring for `Pkg.resolve` is unclear. Originally, we didn't call `Pkg.resolve()` at all and just used `Pkg.instantiate()`, but the Manifest format underwent a backwards-incompatible change a couple years ago forcing it to be included as a hack.

**Comment by tpapp** [comment]
> Current approach: include a Manifest, but try to update it if something is broken.

I am under the impression that it is regenerated unconditionally. I tested with #54 too.

Sorry to be dense, but if you update anyway, then maybe you don't need the very specific versions in the manifest and could live with bounds. Is this correct?

> Something else I haven't thought of?

Maybe not pin, but provide bounds (or enumerate major versions that work) in `Project.toml`, and remove the manifest from the repo.

Given that the package has 2 dependencies + Julia, this may be feasible. Eg LanguageServer.jl gets minor version bumps every 6--12 months or so.

**Comment by non-Jedi** [comment]
> I am under the impression that it is regenerated unconditionally. I tested with [#54](https://github.com/non-Jedi/eglot-jl/pull/54) too.

This isn't the behavior I'm seeing:

```
$ stat -c %y Manifest.toml
2025-01-30 08:46:14.840641589 -0500
$ julia --project=. eglot-jl.jl /home/adam/repos/Loess.jl/
  No Changes to `~/.emacs.d/straight/repos/eglot-jl/Project.toml`
  No Changes to `~/.emacs.d/straight/repos/eglot-jl/Manifest.toml`
[ Info: Environment successfully resolved
┌ Info: Running language server
│   env = "/home/adam/.emacs.d/straight/repos/eglot-jl/Project.toml"
│   src_path = "/home/adam/repos/Loess.jl/"
│   project_path = "/home/adam/repos/Loess.jl"
└   depot_path = ""
^C
[3282] signal 2: Interrupt
... (stacktrace removed as irrelevant)

$ stat -c %y Manifest.toml
2025-01-30 08:46:14.840641589 -0500
```

> Sorry to be dense, but if you update anyway, then maybe you don't need the very specific versions in the manifest and could live with bounds. Is this correct?

I'm open to deleting the Manifest at least for a while. If we get a lot of non-reproducible bug reports, we can always start including it again. Thoughts, @ffevotte?

> Maybe not pin, but provide bounds (or enumerate major versions that work) in `Project.toml`, and remove the manifest from the repo.
> 
> Given that the package has 2 dependencies + Julia, this may be feasible. Eg LanguageServer.jl gets minor version bumps every 6--12 months or so.

To be honest, it's less LanguageServer.jl I'm worried about and would like to pin and more the transitive dependencies which brings the total from 2 to 53.

**Comment by ffevotte** [comment]
> I'm open to deleting the Manifest at least for a while. If we get a lot of non-reproducible bug reports, we can always start including it again. Thoughts, [@ffevotte](https://github.com/ffevotte)?

Yes, I think that would be a useful thing to try: not only is it easy to do, but it will also provide some feedback about the multiplicity of environments people will end up instantiating.

I fear this will not be enough, though. There is for example this issue I encountered last week; it might be better to start a new issue to track this, but here is the short story. I tend to use 1.11 for all projects, and I therefore have an eglot Manifest that is significantly different from the one we ship because it's been instantiated with julia 1.11.3. But for one particular project at work we had to stick to 1.10. I therefore set up a juliaup "override" to automatically use 1.10.8 in the relevant directory. And that made eglot crash for that project, because Pkg wasn't able to automatically re-resolve the environment.

For the time being, this issue can be avoided by customizing `eglot-jl-language-server-project` for eglot to use a separate environment for that project. But this shows that eglot-jl isn't currently designed to work seamlessly in systems where multiple julia versions may be used.

> Something else I haven't thought of?

A category of solutions we haven't considered yet would be to devise an "environment management system" for eglot-jl. By that, I mean a sort of generalization of the new versioned-manifest feature in Julia 1.11. I haven't thought about all details yet, but it could be something along the lines of:

1. before starting the LS, we look at the julia version and try to find a directory named `eglot-env-vX.Y.Z`
2. if such a directory does not exist, we create it, copy the global `Project.toml` in it and instantiate it
3. then we start the LanguageServer in this version-specific environment.

I think this would address most issues we have:
- if we want to ship versioned Manifests, we can (even for julia versions less than 1.11 and even for minor versions)
- if we don't ship a Manifest for a given version, it's not an issue: one will be created automatically. We can also do fancy tricks (like duplicate the `vX.Y.Z-1` environment and start from there to create the `vX.Y.Z` env)
- if at any given time, several julia versions are used on the same system, it's not a problem either: each one of them will use a separate environment

Do you think something like this could work? And would be worth the effort?

**Comment by ffevotte** [comment]
> 5\. Something else I haven't thought of?

Oh, and AFAIU VSCode uses yet another approach: IIUC they vendor all deps in a specific version. I'm not sure about the details of this, but it may be worth investigating.

**Comment by ffevotte** [comment]
> devise an "environment management system" for eglot-jl.

A first draft for this can be found in #57
